tests/federation_encryption_test.go

```diff
@@ -0,0 +1,63 @@
+package tests
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+)
+
+func TestEncryption(t *testing.T) {
+	emptyString := ""
+
+	deployment := Deploy(t, b.BlueprintFederationOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs2", "@bob:hs2")
+
+	encAlice := client.MustCreateEncryptedClient(t, alice)
+	encBob := client.MustCreateEncryptedClient(t, bob)
+
+	// start syncing using mautrix.Client, needed to update the olmMachine store
+	go encAlice.Sync(t)
+	go encBob.Sync(t)
+	defer encAlice.StopSync()
+	defer encBob.StopSync()
+
+	// Create a new room and invite Bob
+	roomID := encAlice.CreateRoom(t, map[string]interface{}{
+		"invite": []string{bob.UserID},
+		"preset": "private_chat",
+	})
+	// Wait for invite and join Bob
+	encBob.MustSyncUntil(t, client.SyncReq{}, client.SyncInvitedTo(bob.UserID, roomID))
+	encBob.JoinRoom(t, roomID, []string{})
+	encAlice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// Enable encryption
+	encAlice.SendEventSynced(t, roomID, b.Event{
+		Type:     "m.room.encryption",
+		StateKey: &emptyString,
+		Sender:   encAlice.UserID,
+		Content: map[string]interface{}{
+			"algorithm": "m.megolm.v1.aes-sha2",
+		},
+	})
+
+	// Send an encrypted text message
+	msgType := "m.text"
+	wantBody := "Hello World!"
+	encAlice.SendEncryptedMessageSynced(t, roomID, msgType, wantBody)
```

Comment:
This hides too much complexity. It's nice that we have a one-liner to send encrypted messages. Let's keep that. _However_, it is hard to see how we can test E2EE more thoroughly with this, as things like uploading/downloading OTKs, syncing device lists, etc is completely hidden from the user. Ideally there would be escape hatches where you can choose to do more complex things, but by default it Just Works.